# Музыка больших городов

Проект по исследованию потребительского поведения пользователей музыкального стримингового сервиса [Яндекс Музыка](https://music.yandex.ru/) — сравнение музыкальных предпочтений пользователей из Москвы и Санкт-Петербурга.

**Статус проекта:** ✅завершен, проверен ревьюером.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Постановка-задачи" data-toc-modified-id="Постановка-задачи-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Постановка задачи</a></span></li><li><span><a href="#Резюме-проекта" data-toc-modified-id="Резюме-проекта-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Резюме проекта</a></span></li><li><span><a href="#Краткое-содержание-работы" data-toc-modified-id="Краткое-содержание-работы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Краткое содержание работы</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек-и-данных" data-toc-modified-id="Импорт-библиотек-и-данных-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Импорт библиотек и данных</a></span></li><li><span><a href="#Обзор-датасета" data-toc-modified-id="Обзор-датасета-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Обзор датасета</a></span></li><li><span><a href="#Выводы-по-разделу-4" data-toc-modified-id="Выводы-по-разделу-4-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Выводы по разделу 4</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Стиль-заголовков-столбцов" data-toc-modified-id="Стиль-заголовков-столбцов-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Стиль заголовков столбцов</a></span></li><li><span><a href="#Пропуски-значений" data-toc-modified-id="Пропуски-значений-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Пропуски значений</a></span></li><li><span><a href="#Дубликаты" data-toc-modified-id="Дубликаты-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Дубликаты</a></span></li><li><span><a href="#Выводы-по-разделу-5" data-toc-modified-id="Выводы-по-разделу-5-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Выводы по разделу 5</a></span></li></ul></li><li><span><a href="#Проверка-гипотез" data-toc-modified-id="Проверка-гипотез-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Проверка гипотез</a></span><ul class="toc-item"><li><span><a href="#Сравнение-поведения-пользователей-двух-столиц" data-toc-modified-id="Сравнение-поведения-пользователей-двух-столиц-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Сравнение поведения пользователей двух столиц</a></span><ul class="toc-item"><li><span><a href="#Предварительные-расчеты" data-toc-modified-id="Предварительные-расчеты-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>Предварительные расчеты</a></span></li><li><span><a href="#Проверка-гипотезы.-Вариант-1-(первоначальный,-от-авторов-курса)" data-toc-modified-id="Проверка-гипотезы.-Вариант-1-(первоначальный,-от-авторов-курса)-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>Проверка гипотезы. Вариант 1 (первоначальный, от авторов курса)</a></span></li><li><span><a href="#Проверка-гипотезы.-Вариант-2-(более-поздний,-собственный)" data-toc-modified-id="Проверка-гипотезы.-Вариант-2-(более-поздний,-собственный)-6.1.3"><span class="toc-item-num">6.1.3&nbsp;&nbsp;</span>Проверка гипотезы. Вариант 2 (более поздний, собственный)</a></span></li><li><span><a href="#Выводы-по-подразделу-6.1" data-toc-modified-id="Выводы-по-подразделу-6.1-6.1.4"><span class="toc-item-num">6.1.4&nbsp;&nbsp;</span>Выводы по подразделу 6.1</a></span></li></ul></li><li><span><a href="#Музыка-в-начале-и-в-конце-недели" data-toc-modified-id="Музыка-в-начале-и-в-конце-недели-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Музыка в начале и в конце недели</a></span><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-6.2.1"><span class="toc-item-num">6.2.1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Функция-для-расчета-топ-10-жанров,-прослушанных-в-заданный-день-и-временной-интервал" data-toc-modified-id="Функция-для-расчета-топ-10-жанров,-прослушанных-в-заданный-день-и-временной-интервал-6.2.2"><span class="toc-item-num">6.2.2&nbsp;&nbsp;</span>Функция для расчета топ-10 жанров, прослушанных в заданный день и временной интервал</a></span></li><li><span><a href="#Проверка-гипотезы.-Вариант-1-(первоначальный,-от-авторов-курса)" data-toc-modified-id="Проверка-гипотезы.-Вариант-1-(первоначальный,-от-авторов-курса)-6.2.3"><span class="toc-item-num">6.2.3&nbsp;&nbsp;</span>Проверка гипотезы. Вариант 1 (первоначальный, от авторов курса)</a></span></li><li><span><a href="#Проверка-гипотезы.-Вариант-2-(более-поздний,-собственный)" data-toc-modified-id="Проверка-гипотезы.-Вариант-2-(более-поздний,-собственный)-6.2.4"><span class="toc-item-num">6.2.4&nbsp;&nbsp;</span>Проверка гипотезы. Вариант 2 (более поздний, собственный)</a></span></li><li><span><a href="#Выводы-по-подразделу-6.2" data-toc-modified-id="Выводы-по-подразделу-6.2-6.2.5"><span class="toc-item-num">6.2.5&nbsp;&nbsp;</span>Выводы по подразделу 6.2</a></span></li></ul></li><li><span><a href="#Жанровые-предпочтения-в-Москве-и-Петербурге" data-toc-modified-id="Жанровые-предпочтения-в-Москве-и-Петербурге-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Жанровые предпочтения в Москве и Петербурге</a></span><ul class="toc-item"><li><span><a href="#Топ-10-жанров-по-Москве" data-toc-modified-id="Топ-10-жанров-по-Москве-6.3.1"><span class="toc-item-num">6.3.1&nbsp;&nbsp;</span>Топ-10 жанров по Москве</a></span></li><li><span><a href="#Топ-10-жанров-по-Петербургу" data-toc-modified-id="Топ-10-жанров-по-Петербургу-6.3.2"><span class="toc-item-num">6.3.2&nbsp;&nbsp;</span>Топ-10 жанров по Петербургу</a></span></li><li><span><a href="#Выводы-по-подразделу-6.3" data-toc-modified-id="Выводы-по-подразделу-6.3-6.3.3"><span class="toc-item-num">6.3.3&nbsp;&nbsp;</span>Выводы по подразделу 6.3</a></span></li></ul></li></ul></li><li><span><a href="#Итоги-исследования" data-toc-modified-id="Итоги-исследования-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Итоги исследования</a></span></li><li><span><a href="#Чек-лист-выполнения-проекта" data-toc-modified-id="Чек-лист-выполнения-проекта-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Чек-лист выполнения проекта</a></span></li></ul></div>

## Постановка задачи
Сравнение Москвы и Петербурга окружено мифами:
- Москва — мегаполис, подчинённый жёсткому ритму рабочей недели;
- Петербург — город своеобразной культуры, непохожий на Москву.

Некоторые мифы отражают действительность. Другие — пустые стереотипы. Бизнес должен отличать первые от вторых, чтобы принимать рациональные решения.

**Цель исследования** — проверить гипотезы о том, как потребительское поведение пользователей Яндекс Музыки различается в зависимости от дня недели, времени суток и города слушателя.

Проверяемые гипотезы:

1. Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.
1. Утром в понедельник в Москве преобладают одни жанры музыки, а в Петербурге — другие. Это верно и для вечера пятницы.
1. Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.

**Задачи исследования:**

1. Проверить исходные данные на ошибки и оценить их влияние на исследование.
1. Определить возможность исправить самые критичные ошибки данных.
1. Проверить предложенные гипотезы.

**Источник данных:** файл `yandex_music_project.csv` с данными о поведении пользователей на платформе Яндекс Музыка.

Описание колонок:
- `userID` — идентификатор пользователя;
- `Track` — название трека;
- `artist` — имя исполнителя;
- `genre` — название жанра;
- `City` — город пользователя;
- `time` — время начала прослушивания;
- `Day` — день недели.


**Навыки и инструменты** `Python`, `pandas`, `.groupby()`,  `.pivot_table()`.

[В начало раздела](#Постановка-задачи) | [К содержанию проекта](#Содержание)

## Резюме проекта
- Из трёх исследованных гипотез первая подтвердилась полностью, вторая и третья — частично.


- Полученные результаты показывают, что во вкусах пользователей Москвы и Петербурга больше общего, чем различий. Если различия в предпочтениях и существуют, на основной массе пользователей они незаметны.


- Пропущенные значения в поле `genre` занимают существенную долю в исходных данных, что ставит под сомнения выводы исследования по второй и третьей гипотезе.


- Из данных одного сервиса не всегда можно сделать вывод о всех жителях города. Сбор данных по другим стриминговым платформам и проверки гипотез статистическими методами в будущем покажут, насколько они достоверны, исходя из имеющихся данных.

[В начало раздела](#Резюме-проекта) | [К содержанию проекта](#Содержание)

## Краткое содержание работы

Исследование проводилось в три этапа:

1. **Обзор данных** — загрузка библиотеки `pandas` и файла с исходными данными, первичное знакомство с данными посредством методов `head()` и `info()`.


2. **Предобработка данных** — устранение выявленных при обзоре недостатков в данных: нарушений в стиле заголовков столбцов, заполнение пропущенных значений, удаление полных дубликатов:
    - все заголовки были приведены к единому стилю методами `df.columns` и `rename()`, а пропущенные значения заменены на 'unknown' при помощи метода `fillna()`;
    - явные дубликаты были обработаны при помощи метода `drop_duplicates()` ;
    - для поиска неявных дубликатов использовались методы `sort_values()` и `unique()`, а для их замены — созданная функция `replace_wrong_genres (wrong_genres, correct_genre)`, использующая метод `replace()`.

</br></br></br>
3. **Проверка гипотез** — получение количественных показателей активности пользователей и проверка исходных гипотез на основании сравнения полученных результатов.

Данный проект был первым в процессе обучения и поэтому не был полностью самостоятельным. Логика и структура проекта, а также способы решения поставленных задач были изначально предложены авторами обучающей программы (в настоящее время этот проект проходится полностью в тренажере платформы без ревью). От учащихся требовалось только правильно применить самые простые методы (вроде `info()`, `value_counts()` и т.п.) и создать функции по предложенной схеме.

Закончив обучение и вернувшись к данному проекту я посчитал нужным улучшить код в некоторых пунктах, поэтому в разделах 6.1 и 6.2 возникли варианты проверок гипотез (первый — первоначальный, от авторов курса, второй — более поздний собственный):

- **в пункте 6.1** вместо [предложенного авторами курса многократного ручного вызова функции](#Проверка-гипотезы.-Вариант-1-(первоначальный,-от-авторов-курса)-6.1.2), подсчитывающей количество прослушиваний по городу в определенный день недели и сведения результатов расчетов в новый датафрейм методом `pd.DataFrame()`, в более позднем варианте [применена сводная таблица](#Проверка-гипотезы.-Вариант-2-(более-поздний,-собственный)-6.1.3) со сглаживанием индекса и настройкой названий столбцов;


- **в пункте 6.2** вместо [предложенного авторами курса многократного ручного вызова функции](#Проверка-гипотезы.-Вариант-1-(первоначальный,-от-авторов-курса)-6.2.3), рассчитывающей рейтинг топ-10 жанров в зависимости от города и временного интервала в течение дня недели, в более позднем варианте [применен цикл с перевбором заранее заданных дней недели и временных интервалов](#Проверка-гипотезы.-Вариант-2-(более-поздний,-собственный)-6.2.4).

[В начало раздела](#Краткое-содержание-работы) | [К содержанию проекта](#Содержание)

## Обзор данных
**Задача:** составить первое представление о данных Яндекс.Музыки.
### Импорт библиотек и данных
Основной инструмент аналитика — `pandas`. Импортируем эту библиотеку.

In [1]:
# Импорт библиотеки pandas
import pandas as pd

Прочитаем файл `yandex_music_project.csv` из папки `/datasets` и сохраним его в переменной `df`.

In [2]:
# Чтение файла с данными и сохранение в df
df = pd.read_csv('datasets/yandex_music_project.csv')

Выведем на экран первые десять строк таблицы.

In [3]:
# Получение первых 10 строк таблицы df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                    Преданная         IMPERVTOR  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE     И вновь продолжается бой               NaN  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

             City        time        Day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
4            Moscow  08:34:34     Monday  
5  Saint-Petersburg  13:09:41     Friday  
6            Moscow  13:00:07  Wednesday  
7            Moscow  20:47:49  Wednesday  
8            Moscow  09:17:40     Friday  
9  Saint-Petersburg  21:20:49  Wednesday

### Обзор датасета
**Задача:** одной командой получить общую информацию о таблице.

In [4]:
# Получение общей информации о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63848 non-null  object
 2   artist    57876 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


**Результаты обзора**
- Всего в исходной таблице содержится **65079** записей, разбитых на **7** столбцов:
    - `userID` — идентификатор пользователя;
    - `Track` — название трека;  
    - `artist` — имя исполнителя;
    - `genre` — название жанра;
    - `City` — город пользователя;
    - `time` — время начала прослушивания;
    - `Day` — день недели.
</br></br>
- Тип данных во всех столбцах — `object`.


- В названиях колонок видны три нарушения стиля:
    1. Строчные буквы сочетаются с прописными.
    2. Встречаются пробелы.
    3. Из названия столбцов не совсем понятно, что за данные хранятся в столбце.
</br></br>
- Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.

### Выводы по разделу 4

- В каждой строке таблицы — данные о прослушанном треке. Часть колонок описывает саму композицию: название, исполнителя и жанр. Остальные данные рассказывают о пользователе: из какого он города, когда он слушал музыку. 


- Предварительно можно утверждать, что, данных достаточно для проверки гипотез. Но встречаются пропуски в данных, а в названиях колонок — расхождения с хорошим стилем.


- Пропуски в данных наблюдаются только в колонках, описывающих саму композицию, что может свидетельствовать об ошибках при извлечении данных о композициях, либо о том, что в системе поля `Track`, `artist` и `genre` являются необязательными и теоретически возможна ситуация, когда все три поля будут пропущены одновременно. *Целесообразно проконсультироваться с разработчиками.*


- С точки зрения постановки задачи наиболее важным является столбец `genre`. Предварительно, он содержит меньше всего пропусков данных.


- Чтобы двигаться дальше, нужно устранить проблемы в данных.

[В начало раздела](#Обзор-данных) | [К содержанию проекта](#Содержание)

## Предобработка данных
**Задачи:**
- исправить стиль в заголовках столбцов;
- исключить пропуски;
- проверить данные на дубликаты.

### Стиль заголовков столбцов
Выведем на экран названия столбцов.

In [5]:
# Перечень названий столбцов таблицы df
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Приведём названия столбцов в соответствие с хорошим стилем:
* несколько слов в названии запишем в «змеином_регистре»,
* все символы сделаем строчными,
* устраним пробелы.

Для этого переименуем колонки с помощью метода `rename()` так:
* `'  userID'` → `'user_id'`;
* `'Track'` → `'track'`;
* `'  City  '` → `'city'`;
* `'Day'` → `'day'`.

In [6]:
# Переименование столбцов
df = df.rename(columns={'  userID': 'user_id', 'Track': 'track', '  City  ': 'city', 'Day': 'day'})

Проверим результат. Для этого ещё раз выведем на экран названия столбцов.

In [7]:
# Проверка результатов - перечень названий столбцов
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

**Вывод:** ✅столбцы переименованы корректно

### Пропуски значений
Посчитаем, сколько в таблице пропущенных значений.

In [8]:
# Подсчёт пропусков
df.isna().sum()

user_id       0
track      1231
artist     7203
genre      1198
city          0
time          0
day           0
dtype: int64

Не все пропущенные значения влияют на исследование. Так в `track` и `artist` пропуски не важны для исследования. Достаточно заменить их явными обозначениями.

Но пропуски в `genre` могут помешать сравнению музыкальных вкусов в Москве и Санкт-Петербурге. На практике было бы правильно установить причину пропусков и восстановить данные. Такой возможности нет в учебном проекте. Придётся:
* заполнить и эти пропуски явными обозначениями,
* оценить, насколько они повредят расчётам.

Заменим пропущенные значения в столбцах `track`, `artist` и `genre` на строку `'unknown'`. Для этого создадим список `columns_to_replace`, переберём его элементы циклом `for` и для каждого столбца выполним замену пропущенных значений.

In [9]:
# Перебор названий столбцов в цикле и замена пропущенных значений на 'unknown'

columns_to_replace = ['track', 'artist', 'genre']
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')

Убедимся, что в таблице не осталось пропусков. Для этого ещё раз посчитаем пропущенные значения.

In [10]:
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

**Вывод:** ✅список `columns_to_replace` сформирован верно, цикл отработал без ошибок.

### Дубликаты
Посчитаем и удалим явные дубликаты в таблице.

In [11]:
# Подсчёт явных дубликатов
df.duplicated().sum()

3826

Вызовем специальный метод `pandas`, чтобы удалить явные дубликаты.

In [12]:
 # Удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index(drop=True)

Ещё раз посчитаем явные дубликаты в таблице — убедимся, что полностью от них избавились:

In [13]:
# Проверка на отсутствие дубликатов
df.duplicated().sum()

0

Избавимся от неявных дубликатов в колонке `genre`. Например, название одного и того же жанра может быть записано немного по-разному. Такие ошибки тоже могут повлиять на результат исследования.


Выведем на экран список уникальных названий жанров, отсортированный в алфавитном порядке. Для этого:
* извлечём нужный столбец датафрейма, 
* применим к нему метод сортировки,
* для отсортированного столбца вызовем метод, который вернёт уникальные значения из столбца.

In [14]:
 # Просмотр уникальных названий жанров
df['genre'].sort_values().unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

Обнаружены следующие неявные дубликаты:
* *hip*,
* *hop*,
* *hip-hop*.

Кроме этого, сортировка уникальных значений столбца `genre` также помогла обнаружить два неявных дубликата, возникших, очевидно, из-за ошибок при вводе:
* *ïîï* — название жанра `pop` на русском языке в кодировке WINDOWS-1251 (это удалось установить с помощью [этого универсального декодера](https://2cyr.com/decode/?lang=ru))
* *электроника* — название жанра `electronic` на русском языке

Чтобы очистить от них таблицу, напишем функцию `replace_wrong_genres()` с двумя параметрами: 
* `wrong_genres` — список дубликатов,
* `correct_genre` — строка с правильным значением.

Функция должна исправить колонку `genre` в таблице `df`: заменить каждое значение из списка `wrong_genres` на значение из `correct_genre`.

In [15]:
# Функция для замены неявных дубликатов

def replace_wrong_genres (wrong_genres, correct_genre): # на вход функции подаются список дубликатов и строка с правильным значением
    for wrong_genre in wrong_genres: # перебираем дубликаты в списке wrong_genres
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre) # и для каждого неправильного названия жанра вызываем метод replace()

Создадим словарь `genres_to_replace` в котором каждому правильному названию жанра будут сопоставлены варианты, которые должны быть заменены: вместо `hip`, `hop` и `hip-hop` должно быть значение `hiphop`, также `ïîï` должно быть заменено на `pop`, а `электроника` на `electronic`. После этого, в цикле по элементам этого словаря вызовем функцию `replace_wrong_genres()` и передадим ей такие аргументы, чтобы она устранила неявные дубликаты.

In [16]:
# Устранение неявных дубликатов

genres_to_replace = {
    'hiphop' : ['hip', 'hop', 'hip-hop'],
    'pop' : ['ïîï'],
    'electronic' : ['электроника']
}

for correct_value in genres_to_replace:
    replace_wrong_genres(genres_to_replace[correct_value], correct_value)

Проверим корректность устранения неявных дубликатов. Выведем отсортированный список уникальных значений столбца `genre`.

In [17]:
df['genre'].sort_values().unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

**Вывод:** ✅неявные дубликаты удалены корректно.

### Выводы по разделу 5
- Предобработка обнаружила три проблемы в данных:
    - нарушения в стиле заголовков,
    - пропущенные значения,
    - дубликаты — явные и неявные.
</br></br>
- Мы исправили заголовки, чтобы упростить работу с таблицей.


- Без явных дубликатов исследование станет более точным.


- Пропущенные значения мы заменили на `'unknown'`. Далее предстоит проверить, не повредят ли исследованию пропуски в колонке `genre`.

Теперь можно перейти к проверке гипотез. 

[В начало раздела](#Предобработка-данных) | [К содержанию проекта](#Содержание)

## Проверка гипотез

### Сравнение поведения пользователей двух столиц
**Задача:** первая гипотеза утверждает, что пользователи по-разному слушают музыку в Москве и Санкт-Петербурге. Необходимо проверить это предположение по данным о трёх днях недели — понедельнике, среде и пятнице. Для этого требуется:
* разделить пользователей Москвы и Санкт-Петербурга;
* сравнить, сколько треков послушала каждая группа пользователей в понедельник, среду и пятницу.

#### Предварительные расчеты

Сначала выполним каждый из расчётов по отдельности. Оценим активность пользователей в каждом городе. Сгруппируем данные по городу и посчитаем прослушивания в каждой группе.

In [18]:
# Подсчёт прослушиваний в каждом городе
df.groupby('city')['genre'].count()

city
Moscow              42741
Saint-Petersburg    18512
Name: genre, dtype: int64

**Промежуточный выводы:**
- в Москве прослушиваний больше, чем в Петербурге;
- из этого не следует, что московские пользователи чаще слушают музыку — просто самих пользователей в Москве больше.

Теперь сгруппируем данные по дню недели и подсчитаем прослушивания в понедельник, среду и пятницу. Учтём, что в данных есть информация о прослушиваниях только за эти дни.

In [19]:
# Подсчёт прослушиваний в каждый из дней
df.groupby('day')['day'].count()

day
Friday       21840
Monday       21354
Wednesday    18059
Name: day, dtype: int64

**Промежуточный выводы:**
- в среднем пользователи из двух городов менее активны по средам;
- картина может измениться, если рассмотреть каждый город в отдельности.

#### Проверка гипотезы. Вариант 1 (первоначальный, от авторов курса)
Увидели, как работает группировка по городу и по дням недели. Теперь напишем функцию, которая объединит два эти расчёта.

Создадим функцию `number_tracks()`, которая посчитает прослушивания для заданного дня и города. Ей понадобятся два параметра:
* день недели,
* название города.

В функции сохраним в переменную строки исходной таблицы, у которых значение:
  * в колонке `day` равно параметру `day`,
  * в колонке `city` равно параметру `city`.

Для этого применим последовательную фильтрацию с логической индексацией.

Затем посчитаем значения в столбце `user_id` получившейся таблицы. Результат сохраним в новую переменную. Вернём эту переменную из функции.

In [20]:
# Создание функции number_tracks()

# Объявляется функция с двумя параметрами: day, city.
def number_tracks(day, city):
    
# В переменной track_list сохраняются те строки таблицы df, для которых 
# значение в столбце 'day' равно параметру day и одновременно значение
# в столбце 'city' равно параметру city (используйте последовательную
# фильтрацию с помощью логической индексации).
    
    track_list = df[df['day'] == day]
    track_list = track_list[track_list['city'] == city]


# В переменной track_list_count сохраняется число значений столбца 'user_id',
# рассчитанное методом count() для таблицы track_list.
    
    track_list_count = track_list['user_id'].count()

# Функция возвращает число - значение track_list_count.
    
    return track_list_count
                                  
# Функция для подсчёта прослушиваний для конкретного города и дня.
# С помощью последовательной фильтрации с логической индексацией она 
# сначала получит из исходной таблицы строки с нужным днём,
# затем из результата отфильтрует строки с нужным городом,
# методом count() посчитает количество значений в колонке user_id. 
# Это количество функция вернёт в качестве результата

Вызовем `number_tracks()` шесть раз, меняя значение параметров — так, чтобы получить данные для каждого города в каждый из трёх дней.

In [21]:
# Количество прослушиваний в Москве по понедельникам
number_tracks('Monday','Moscow')

15740

In [22]:
# Количество прослушиваний в Санкт-Петербурге по понедельникам
number_tracks('Monday','Saint-Petersburg')

5614

In [23]:
# Количество прослушиваний в Москве по средам
number_tracks('Wednesday','Moscow')

11056

In [24]:
# Количество прослушиваний в Санкт-Петербурге по средам
number_tracks('Wednesday','Saint-Petersburg')

7003

In [25]:
# Количество прослушиваний в Москве по пятницам
number_tracks('Friday','Moscow')

15945

In [26]:
# Количество прослушиваний в Санкт-Петербурге по пятницам
number_tracks('Friday','Saint-Petersburg')

5895

Создадим c помощью конструктора `pd.DataFrame` таблицу, где
* названия колонок — `['City', 'Monday', 'Wednesday', 'Friday']`;
* данные — результаты, которые вы получили с помощью `number_tracks`.

Добавим также для наглядности столбец 'Total' в котором укажем общее количество прослушиваний для каждого города (сумму столбцов `Monday`, `Wednesday`, `Friday`)

In [27]:
# Таблица с результатами
data = [['Moscow','|', 42741,'|', 15740, 11056, 15945],
       ['Saint-Petersburg','|', 18512,'|', 5614, 7003, 5895]]

columns = ['City','','Total','','Monday','Wednesday','Friday']
pd.DataFrame(data, columns=columns)

City     Total     Monday  Wednesday  Friday
0            Moscow  |  42741  |   15740      11056   15945
1  Saint-Petersburg  |  18512  |    5614       7003    5895

#### Проверка гипотезы. Вариант 2 (более поздний, собственный)
Построим сводную таблицу

In [28]:
# Формируем сводную таблицу
pivot_table = df.pivot_table(index=['city'],
                             columns='day',
                             values='user_id',
                             aggfunc='count',
                             fill_value=0,
                             margins=True,
                             margins_name='Total')

# Обрабатываем заголовки столбцов с сводной таблице
pivot_table.columns = [''.join(str(s).strip() for s in col if s) for col in pivot_table.columns]

# Сбрасываем индекс сводной таблицы
pivot_table.reset_index(inplace=True)

# Упорядочиваем таблицу в порядке дней недели
pivot_table = pivot_table[['city', 'Monday', 'Wednesday', 'Friday', 'Total']]

pivot_table

city  Monday  Wednesday  Friday  Total
0            Moscow   15740      11056   15945  42741
1  Saint-Petersburg    5614       7003    5895  18512
2             Total   21354      18059   21840  61253

#### Выводы по подразделу 6.1
- Данные показывают разницу поведения пользователей:
    - в Москве пик прослушиваний приходится на понедельник и пятницу, а в среду заметен спад.
    - в Петербурге, наоборот, больше слушают музыку по средам. Активность в понедельник и пятницу здесь почти в равной мере уступает среде.
</br></br>
- Значит, данные говорят в пользу первой гипотезы.

### Музыка в начале и в конце недели
**Задача:** проверить гипотезу о том, что утром в понедельник в Москве преобладают одни жанры, а в Петербурге — другие, а так же и вечером пятницы преобладают разные жанры — в зависимости от города.
#### Подготовка данных
Сохраним таблицы с данными в две переменные:
* по Москве — в `moscow_general`;
* по Санкт-Петербургу — в `spb_general`.

In [29]:
# Получение таблицы moscow_general из тех строк таблицы df, 
# для которых значение в столбце 'city' равно 'Moscow'

moscow_general = df[df['city'] == 'Moscow']
display(moscow_general.head())

# Получение таблицы spb_general из тех строк таблицы df,
# для которых значение в столбце 'city' равно 'Saint-Petersburg'

spb_general = df[df['city'] == 'Saint-Petersburg']
spb_general.head()

user_id                        track            artist   genre    city  \
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock  Moscow   
4  E2DC1FAE                  Soul People        Space Echo   dance  Moscow   
6  4CB90AA5                         True      Roman Messer   dance  Moscow   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance  Moscow   
8  8FA1D3BE     И вновь продолжается бой           unknown  ruspop  Moscow   

       time        day  
1  14:07:09     Friday  
4  08:34:34     Monday  
6  13:00:07  Wednesday  
7  20:47:49  Wednesday  
8  09:17:40     Friday

user_id                  track            artist   genre  \
0  FFB692EC      Kamigata To Boots  The Mass Missile    rock   
2    20EC38      Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9  Dragons in the Sunset        Fire + Ice    folk   
5  842029A1              Преданная         IMPERVTOR  rusrap   
9  E772D5C0              Pessimist           unknown   dance   

               city      time        day  
0  Saint-Petersburg  20:28:33  Wednesday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
5  Saint-Petersburg  13:09:41     Friday  
9  Saint-Petersburg  21:20:49  Wednesday

#### Функция для расчета топ-10 жанров, прослушанных в заданный день и временной интервал
Создадим функцию `genre_weekday()` с четырьмя параметрами:
* таблица (датафрейм) с данными,
* день недели,
* начальная временная метка в формате 'hh:mm', 
* последняя временная метка в формате 'hh:mm'.

Функция должна вернуть информацию о топ-10 жанров тех треков, которые прослушивали в указанный день, в промежутке между двумя отметками времени.

In [30]:
# Объявление функции genre_weekday() с параметрами table, day, time1, time2,
# которая возвращает информацию о самых популярных жанрах в указанный день в
# заданное время:
# 1) в переменную genre_df сохраняются те строки переданного датафрейма table, для
#    которых одновременно:
#    - значение в столбце day равно значению аргумента day
#    - значение в столбце time больше значения аргумента time1
#    - значение в столбце time меньше значения аргумента time2
#    Используйте последовательную фильтрацию с помощью логической индексации.
# 2) сгруппировать датафрейм genre_df по столбцу genre, взять один из его
#    столбцов и посчитать методом count() количество записей для каждого из
#    присутствующих жанров, получившийся Series записать в переменную
#    genre_df_count
# 3) отсортировать genre_df_count по убыванию встречаемости и сохранить
#    в переменную genre_df_sorted
# 4) вернуть Series из 10 первых значений genre_df_sorted, это будут топ-10
#    популярных жанров (в указанный день, в заданное время)

def genre_weekday(table, day, time1, time2):
    genre_df = table[table['day'] == day]
    genre_df = genre_df[genre_df['time'] > time1]
    genre_df = genre_df[genre_df['time'] < time2]
    genre_df_count = genre_df.groupby('genre')['user_id'].count()
    genre_df_sorted = genre_df_count.sort_values(ascending=False)
    return genre_df_sorted.head(10)

#### Проверка гипотезы. Вариант 1 (первоначальный, от авторов курса)
Сравните результаты функции `genre_weekday()` для Москвы и Санкт-Петербурга в понедельник утром (с 7:00 до 11:00) и в пятницу вечером (с 17:00 до 23:00):

In [31]:
# Вызов функции для утра понедельника в Москве (вместо df — таблица moscow_general)
# объекты, хранящие время, являются строками и сравниваются как строки

genre_weekday(moscow_general, 'Monday', '07:00', '11:00')

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
Name: user_id, dtype: int64

In [32]:
# Вызов функции для утра понедельника в Петербурге (вместо df — таблица spb_general)

genre_weekday(spb_general, 'Monday', '07:00', '11:00')

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
Name: user_id, dtype: int64

In [33]:
# Вызов функции для вечера пятницы в Москве

genre_weekday(moscow_general, 'Friday', '17:00', '23:00')

genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
alternative    163
classical      163
rusrap         142
Name: user_id, dtype: int64

In [34]:
# Вызов функции для вечера пятницы в Петербурге

genre_weekday(spb_general, 'Friday', '17:00', '23:00')

genre
pop            256
electronic     216
rock           216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
Name: user_id, dtype: int64

#### Проверка гипотезы. Вариант 2 (более поздний, собственный)
Зададим интервалы для которых нужно получить выборки и организуем циклы для перебора возможных вариантов выборок.

In [35]:
dt_interval = [('Monday', '07:00', '11:00'), ('Friday', '17:00', '23:00')]

for day_time in dt_interval:
    day, start, end = day_time
    for city_df in [spb_general, moscow_general]:
        print('Город: {} День: {} Время: {} - {}'.format(city_df['city'].unique()[0], day, start, end))
        print(genre_weekday(city_df, day, start, end))
        print('\n')

Город: Saint-Petersburg День: Monday Время: 07:00 - 11:00
genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
Name: user_id, dtype: int64


Город: Moscow День: Monday Время: 07:00 - 11:00
genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
Name: user_id, dtype: int64


Город: Saint-Petersburg День: Friday Время: 17:00 - 23:00
genre
pop            256
electronic     216
rock           216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
Name: user_id, dtype: int64


Город: Moscow День: Friday Время: 17:00 - 23:00
genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world    

#### Выводы по подразделу 6.2
- Если сравнить топ-10 жанров в понедельник утром, можно сделать такие выводы:
    1. В Москве и Петербурге слушают похожую музыку. Единственное отличие — в московский рейтинг вошёл жанр “world”, а в петербургский — джаз и классика.
    2. В Москве пропущенных значений оказалось так много, что значение `'unknown'` заняло десятое место среди самых популярных жанров. Значит, пропущенные значения занимают существенную долю в данных и угрожают достоверности исследования.
</br></br>
- Вечер пятницы не меняет эту картину. Некоторые жанры поднимаются немного выше, другие спускаются, но в целом топ-10 остаётся тем же самым.


- Таким образом, вторая гипотеза подтвердилась лишь частично:
    * Пользователи слушают похожую музыку в начале недели и в конце.
    * Разница между Москвой и Петербургом не слишком выражена. В Москве чаще слушают русскую популярную музыку, в Петербурге — джаз.
</br></br>
- Однако пропуски в данных ставят под сомнение этот результат. В Москве их так много, что рейтинг топ-10 мог бы выглядеть иначе, если бы не утерянные  данные о жанрах.

### Жанровые предпочтения в Москве и Петербурге

**Задача:** проверить гипотезу: Петербург — столица рэпа, музыку этого жанра там слушают чаще, чем в Москве, а Москва — город контрастов, в котором, тем не менее, преобладает поп-музыка.

#### Топ-10 жанров по Москве

Сгруппируем таблицу `moscow_general` по жанру и посчитаем прослушивания треков каждого жанра методом `count()`. Затем отсортируем результат в порядке убывания и сохраним его в таблице `moscow_genres`.

In [36]:
# Одной строкой: группировка таблицы moscow_general по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в moscow_genres

moscow_genres = moscow_general.groupby('genre')['genre'].count().sort_values(ascending=False)

# Просмотр первых 10 строк moscow_genres
moscow_genres.head(10)

genre
pop            5893
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64

#### Топ-10 жанров по Петербургу
Сгруппируем таблицу `spb_general` по жанру. Посчитаем прослушивания треков каждого жанра. Результат отсортируем в порядке убывания и сохраним в таблице `spb_genres`:

In [37]:
# Группировка таблицы spb_general
spb_genres = spb_general.groupby('genre')['genre'].count().sort_values(ascending=False)

# Просмотр первых 10 строк spb_genres
spb_genres.head(10)

genre
pop            2431
dance          1932
rock           1879
electronic     1737
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64

#### Выводы по подразделу 6.3
Гипотеза частично подтвердилась:
* поп-музыка — самый популярный жанр в Москве, как и предполагала гипотеза. Более того, в топ-10 жанров встречается близкий жанр — русская популярная музыка.
* вопреки ожиданиям, рэп одинаково популярен в Москве и Петербурге.

[В начало раздела](#Проверка-гипотез) | [К содержанию проекта](#Содержание)

## Итоги исследования
Мы проверили три гипотезы и установили:

1. День недели по-разному влияет на активность пользователей в Москве и Петербурге. 

**Таким образом, первая гипотеза полностью подтвердилась.**

2. Музыкальные предпочтения не сильно меняются в течение недели — будь то Москва или Петербург. Небольшие различия заметны в начале недели, по понедельникам:
* в Москве слушают музыку жанра «world»,
* в Петербурге — джаз и классику.

**Таким образом, вторая гипотеза подтвердилась лишь отчасти. Этот результат мог оказаться иным, если бы не пропуски в данных.**

3. Во вкусах пользователей Москвы и Петербурга больше общего чем различий. Вопреки ожиданиям, предпочтения жанров в Петербурге напоминают московские.

**Третья гипотеза не подтвердилась. Если различия в предпочтениях и существуют, на основной массе пользователей они незаметны.**

[В начало раздела](#Итоги-исследования) | [К содержанию проекта](#Содержание)

## Чек-лист выполнения проекта

**Шаг 1. Загрузка данных и изучение общей информации**
   - [x] Открыт файл с исходными данными
   - [x] Файл изучен (выведены первые строки, метод `info()`)
   - [x] Сделан вывод о необходимости предобработки данных

</br></br>
**Шаг 2. Предобработка данных**
   - [x] Заголовки столбцов приведены к хорошему стилю
   - [x] Подсчитаны и заполнены «заглушкой» пропуски в данных 
   - [x] Удалены полные дубликаты
   - [x] Исправлены неявные дубликаты    

</br></br>
**Шаг 3. Проверка гипотез**
   - [x] Проверена первая гипотеза
   - [x] Проверена вторая гипотеза
   - [x] Проверена третья гипотеза
   - [x] Сделан общий вывод по исследованию

</br></br>
[К содержанию проекта](#Содержание)